### PR TITLE
Implement save buttons for metadata

### PR DESCRIPTION
**Pull request by pmiddend**
This PR changes the way metadata is updated in the Admin UI. Instead of the metadata fields being updated *immediately*, there’s a “Save metadata” button at the bottom to bulk-save the metadata. This is consistent with the already-existing bulk update metadata modal (and this modal was also adapted).

Editing a metadata field now doesn’t produce a little checkmark at the right, but a red “is dirty” symbol that I picked at random.

Also included in this PR is the ability to trigger a workflow on all affected events. Meaning: all events in the series in case of a series metadata change, a single event in case of the normal “Edit event” modal or all selected events in case of the “Bulk edit events” modal.

Editing metadata and then trying to exit the modal results in a prompt to confirm unsaved changes.

Also fixed is the “is a workflow running on the event” detection, which behaved weirdly.

Sneakily, this also introduces a small style change to the Javascript DI mechanism, adding more newlines. This hopefully makes merge conflicts easier to resolve and is a bit easier to read.

This work is sponsored by the University of Vienna as part of the 2019/2020 crowdfunding campaign.

### CODE REVIEW

**Comment by pmiddend** [comment]
Note that @mtneug brought up a change that's still missing from this PR: the ability to auto-trigger workflows from the External API. It's not ready to merge just yet.

**Comment by gregorydlogan** [comment]
@pmiddend Did the auto-trigger from the external API get finished?  If not, I'm assuming it's not forthcoming at this point?

**Comment by pmiddend** [comment]
@gregorydlogan Oh damn. Yes, the auto-trigger stuff is finished, I just didn't edit/delete my comment above.

File: modules/external-api/src/main/java/org/opencastproject/external/endpoint/EventsEndpoint.java [hunk 1]
```diff
@@ -421,6 +457,17 @@ public Response deleteEvent(@HeaderParam("Accept") String acceptHeader, @PathPar
     return Response.noContent().build();
   }
 
+  private boolean startChangeWorkflow(final Collection<String> mpIds)
```
Comment (KatrinIhler) [inline modules/external-api/src/main/java/org/opencastproject/external/endpoint/EventsEndpoint.java:462]: In this endpoint, this only ever takes a single id, so why not simply take a String?

File: modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/SeriesEndpoint.java [hunk 2]
```diff
@@ -398,11 +437,26 @@ public Response updateSeriesMetadata(@PathParam("seriesId") String seriesID,
...
+          securityService.getUser()).withSeriesId(seriesID);
+        final SearchResultItem<Event>[] eventsInSerie = searchIndex.getByQuery(eventInSeriesQuery).getItems();
+        final WorkflowDefinition wfd = this.workflowService.getWorkflowDefinitionById(this.metadataChangeWorkflow);
+        final Workflows workflows = new Workflows(this.assetManager, this.workflowService);
+        final ConfiguredWorkflow workflow = workflow(wfd, this.metadataChangeWorkflowConfig);
+        final Set<String> mpIds = Arrays.stream(eventsInSerie).map(e -> e.getSource().getIdentifier()).collect(
+          Collectors.toSet());
+        final List<WorkflowInstance> partialResult = workflows.applyWorkflowToLatestVersion(mpIds, workflow).toList();
+        if (!partialResult.isEmpty())
+          metadataList.setLocked(MetadataList.Locked.WORKFLOW_RUNNING);
```
Comment (KatrinIhler) [inline modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/SeriesEndpoint.java:452]: I'm wondering whether locking series metadata even works?

File: modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/AbstractEventEndpoint.java [hunk 3]
```diff
@@ -1448,13 +1472,29 @@ public Response updateEventMetadata(@PathParam("eventId") String id, @FormParam(
       return notFound("Cannot find an event with id '%s'.", id);
 
     try {
-      MetadataList metadataList = getIndexService().updateAllEventMetadata(id, metadataJSON, getIndex());
+      final MetadataList metadataList = getIndexService().updateAllEventMetadata(id, metadataJSON, getIndex());
+      if (this.metadataChangeWorkflow != null) {
+        final boolean workflowsStarted = startChangeWorkflow(Collections.singleton(id));
+        if (workflowsStarted) {
```
Comment (KatrinIhler) [inline modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/AbstractEventEndpoint.java:1482]: This condition doesn't look right. Shouldn't it be !workflowsStarted?

**Comment by KatrinIhler** [comment]
@pmiddend Conflict!

**Review by KatrinIhler — CHANGES_REQUESTED** [review]
I've tested this and I noticed some weird behavior. Examples:

- "Dirty" symbol in event-details doesn't show up after save was hit once.
- "Dirty" symbols in series-details don't disappear after save was hit.

- When using extended metadata, the event-details modal appears to double for a moment when hitting save. (???)
- Saving extended metadata doesn't seem to work for series.

- I can hit the save button of the series again even if I just started a workflow, leading to an error. This was not supposed to work, right?

**Comment by gregorydlogan** [comment]
@KatrinIhler I see the force push, are your issues resolved?

**Comment by KatrinIhler** [comment]
@gregorydlogan Thanks for the reminder, this completely fell of my radar with everything that's been going on. Hopefully I'll get this done sometime this month...

File: etc/org.opencastproject.adminui.endpoint.OsgiEventEndpoint.cfg [hunk 4]
```diff
@@ -32,3 +32,8 @@ preview.subtype=preview
 # to in the events tab.
 # Default: false
 # eventsTab.onlyEventsWithWriteAccess=false
+
+# Which workflow should be started when metadata properties change?
+# org.opencastproject.admin.ui.event-metadata-change-workflow=republish-metadata
```
Comment (KatrinIhler) [inline etc/org.opencastproject.adminui.endpoint.OsgiEventEndpoint.cfg:37]: TODO: Is this the default? Or is no workflow started by default?

File: modules/admin-ui-frontend/app/scripts/shared/partials/modals/series-details.html [hunk 5]
```diff
@@ -93,6 +93,14 @@ <h2 translate="EVENTS.SERIES.DETAILS.HEADER" translate-values="{{titleParams}}">
           </div>
         </div>
       </div>
+      <footer>
+        <a
+          class="submit"
+          ng-class="{disabled: metadata.locked}"
+          ng-click="metadataSave(seriesCatalog)">
```
Comment (KatrinIhler) [inline modules/admin-ui-frontend/app/scripts/shared/partials/modals/series-details.html:100]: Why does it only get a parameter here?

File: etc/org.opencastproject.adminui.endpoint.OsgiEventEndpoint.cfg [hunk 6]
```diff
@@ -32,3 +32,8 @@ preview.subtype=preview
...
+# Configuration for metadata change workflow
+# org.opencastproject.admin.ui.event-metadata-change-workflow-config={}
```
Comment (KatrinIhler) [inline etc/org.opencastproject.adminui.endpoint.OsgiEventEndpoint.cfg:39]: No newline at end of file

File: etc/org.opencastproject.adminui.endpoint.SeriesEndpoint.cfg [hunk 7]
```diff
@@ -14,3 +14,8 @@
 # to in the events filter.
 # Default: false
 # eventsFilter.onlySeriesWithWriteAccess=false
+
+# Which workflow should be started when metadata properties change?
+# org.opencastproject.admin.ui.series-metadata-change-workflow=republish-metadata
+# Configuration for metadata change workflow
+# org.opencastproject.admin.ui.series-metadata-change-workflow-config={}
```
Comment (KatrinIhler) [inline etc/org.opencastproject.adminui.endpoint.SeriesEndpoint.cfg:21]: See OsgiEventEndpoint.cfg

File: etc/org.opencastproject.adminui.endpoint.SeriesEndpoint.cfg [hunk 8]
```diff
@@ -14,3 +14,8 @@
 # to in the events filter.
 # Default: false
 # eventsFilter.onlySeriesWithWriteAccess=false
+
+# Which workflow should be started when metadata properties change?
+# org.opencastproject.admin.ui.series-metadata-change-workflow=republish-metadata
```
Comment (KatrinIhler) [inline etc/org.opencastproject.adminui.endpoint.SeriesEndpoint.cfg:19]: See OsgiEventEndpoint.cfg

File: etc/org.opencastproject.external.endpoint.EventsEndpoint.cfg [hunk 9]
```diff
@@ -4,6 +4,11 @@
 # time to access the content.
 # url.signing.expires.seconds=7200
 
+# Which workflow should be started when metadata properties change?
+# metadata-change-workflow=republish-metadata
```
Comment (KatrinIhler) [inline etc/org.opencastproject.external.endpoint.EventsEndpoint.cfg:8]: See OsgiEventEndpoint.cfg

File: etc/org.opencastproject.external.endpoint.SeriesEndpoint.cfg [hunk 10]
```diff
@@ -0,0 +1,4 @@
+# Which workflow should be started when metadata properties change?
+# metadata-change-workflow=republish-metadata
```
Comment (KatrinIhler) [inline etc/org.opencastproject.external.endpoint.SeriesEndpoint.cfg:2]: See OsgiEventEndpoint.cfg

File: etc/org.opencastproject.external.endpoint.SeriesEndpoint.cfg [hunk 11]
```diff
@@ -0,0 +1,4 @@
...
+# Configuration for metadata change workflow
+# metadata-change-workflow-config={}
```
Comment (KatrinIhler) [inline etc/org.opencastproject.external.endpoint.SeriesEndpoint.cfg:4]: See OsgiEventEndpoint.cfg

File: modules/admin-ui-frontend/app/scripts/modules/events/controllers/eventController.js [hunk 12]
```diff
@@ -23,18 +23,91 @@
...
+    $sce,
+    $timeout,
+    EventHelperService,
+    UploadAssetOptions,
+    EventUploadAssetResource,
+    Table,
+    SchedulingHelperService,
+    StatisticsReusable,
+    Modal,
+    $translate) {
```
Comment (KatrinIhler) [inline modules/admin-ui-frontend/app/scripts/modules/events/controllers/eventController.js:86]: I'm really not a fan of spacing the dependencies out like this.

File: modules/admin-ui-frontend/app/scripts/modules/events/controllers/eventController.js [hunk 13]
```diff
@@ -23,18 +23,91 @@
...
+
+    $translate('EVENTS.SERIES.DETAILS.WARNING_UNSAVED').then(function (translation) {
+      window.unloadConfirmMsg = translation;
```
Comment (KatrinIhler) [inline modules/admin-ui-frontend/app/scripts/modules/events/controllers/eventController.js:89]: Why not in HTML?

File: modules/admin-ui-frontend/app/scripts/modules/events/controllers/eventController.js [hunk 14]
```diff
@@ -23,18 +23,91 @@
...
+    }).catch(angular.noop);
+
+    var dirtyFieldsPresent = function() {
+      return $scope.metadata.entries.some(
```
Comment (KatrinIhler) [inline modules/admin-ui-frontend/app/scripts/modules/events/controllers/eventController.js:93]: Why check both?

File: modules/admin-ui-frontend/test/test/unit/shared/resources/eventMetadataResourceSpec.js [hunk 15]
```diff
@@ -29,9 +29,9 @@ describe('Event Metadata API Resource', function () {
 
     describe('#save', function () {
         it('sends an array of metadata', function () {
-            var metadataRequest = {
+            var metadataRequest = [{
                 fields: [{ id: 'title' }, { id: 'series' }]
-            };
+            }];
```
Comment (KatrinIhler) [inline modules/admin-ui-frontend/test/test/unit/shared/resources/eventMetadataResourceSpec.js:34]: Why this change?

File: modules/admin-ui-frontend/resources/public/org/opencastproject/adminui/languages/lang-en_US.json [hunk 16]
```diff
@@ -1014,7 +1015,8 @@
          }
        },
        "DETAILS": {
-          "HEADER": "Series details - {{resourceId}}",
+         "HEADER": "Series details - {{resourceId}}",
+         "WARNING_UNSAVED": "Warning, there are unsaved changes. Really quit?",
```
Comment (KatrinIhler) [inline modules/admin-ui-frontend/resources/public/org/opencastproject/adminui/languages/lang-en_US.json:1019]: Right spot?

File: modules/external-api/src/main/java/org/opencastproject/external/endpoint/SeriesEndpoint.java [hunk 17]
```diff
@@ -137,6 +157,34 @@
   private IndexService indexService;
   private SecurityService securityService;
   private SeriesService seriesService;
+  private String metadataChangeWorkflow;
```
Comment (KatrinIhler) [inline modules/external-api/src/main/java/org/opencastproject/external/endpoint/SeriesEndpoint.java:160]: TODO: test external API changes

These changes also require a new external API verson and this should only work for the newer version. 
Also is this documented?

File: modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/AbstractEventEndpoint.java [hunk 18]
```diff
@@ -1452,13 +1476,29 @@ public Response updateEventMetadata(@PathParam("eventId") String id, @FormParam(
...
+          metadataList.setLocked(Locked.WORKFLOW_RUNNING);
+        }
+      }
       return okJson(metadataList.toJSON());
     } catch (IllegalArgumentException e) {
       return badRequest(String.format("Event %s metadata can't be updated.: %s", id, e.getMessage()));
     }
   }
 
+  private boolean startChangeWorkflow(final Collection<String> mpIds)
```
Comment (KatrinIhler) [inline modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/AbstractEventEndpoint.java:1494]: Why does this take a list?

**Review by KatrinIhler — CHANGES_REQUESTED** [review]
Just realized that I forgot to post my review, whoops. Some comments are also works in progress, but I'm gonna leave it as is right now.

The gist of it is: This still doesn't work properly, and since Plapadoo obviously can't fix this, we as a community should think about how we can pour some time into this to fix it, because generally this is something we want.

**Comment by KatrinIhler** [comment]
Since this still requires some work, I converted this to a draft for now.

**Comment by lkiesow** [comment]
This is superseded by pull request #3176

**Pull request closed without merging**